### PR TITLE
Rerender only when data for given reduxes has changed

### PR DESCRIPTION
**Pull request by rawagner**
`ConnectToState` updates children everytime `k8s` map is changed, which means that even if watched resource isn't changed, but something else in `k8s` map changed, we unnecessarily rerender child component. For example, we have two components - `A` and `B`. `A` is watching all pods, `B` is watching all nodes. Pods are updated, `k8s` map is changed and `A` and `B` are rerendered. but `B` doesnt have to be.

To optimize rerenders we can check only parts of `k8s` map and rerender only when the parts we are interested in have changed.

Or maybe I misread something :)

@spadgett

### CODE REVIEW

File: frontend/public/components/utils/firehose.jsx [hunk 1]
```diff
@@ -71,9 +71,20 @@ const worstError = errors => {
   return worst;
 };
 
+const mapStateToProps = ({k8s}) => ({
+  k8s,
+});
+
+const propsAreEqual = (prevProps, nextProps) => {
+  if (nextProps.children === prevProps.children && nextProps.reduxes === prevProps.reduxes) {
```
Comment (christianvogt) [inline frontend/public/components/utils/firehose.jsx]: There are several other props missing from this check: `filters, loaded, loadError, resources`

Comment (rawagner) [reply frontend/public/components/utils/firehose.jsx]: how so ? `ConnectToState` has only 3 props - `children`, `reduxes` (both from Firehose) and `k8s` (from mapStateToProps). Am I wrong ?

Comment (christianvogt) [reply frontend/public/components/utils/firehose.jsx]: Sorry you're right. Not sure how I came to that conclusion before.

**Comment by christianvogt** [comment]
@rawagner i haven't tried locally but it would seem that if `k8s` changes then `Firehose` would re-render and create a new instance of `children`, due to `inject` in its `render` function, which would be passed down to `ConnectToState` and still bypass your `propsAreEqual` check.
What scenario did you try in the app to reproduce this? Otherwise will need to contrive something to test. Would be great if we had some unit tests here :)

**Comment by spadgett** [comment]
I'd suggest we push this to 4.3 to give it more soak time.

**Comment by rawagner** [comment]
> @rawagner i haven't tried locally but it would seem that if `k8s` changes then `Firehose` would re-render 

not necessarily, `Firehose` takes `k8sModels`, `loaded` and `inFlight` from `stateToProps` but there is `areStatePropsEqual` function which actually compares these props to know if it should rerender. So even if `k8s` changes (ie some k8s resource changed) but `models` and `k8s.getIn(['RESOURCES', 'inFlight'])` is the same `Firehose` wont rerender, but `ConnectToState` would rerender. With this change `ConnectToState` additionally checks if a resource we are interested in changed.

I've tried with and without this change on Dashboards Page where a lot of resources are fetched via various cards (all pods, all events etc..) and it reduced the amount of rerenders

**Comment by christianvogt** [comment]
Tried out the changes. While sitting on the dashboard page it almost looks like the same cards are being updated but visually nothing has changed as before applying the fix.
I put conditional breakpoints in the new code and indeed it looks to be working and preventing some re-renders. There are also, as mentioned above, the checks for children are failing and component re-renders are triggered although visually there are no changes.

Seems to me this is an improvement but still more work could be done.

Do you mind posting a specific scenario where you compare the before and after?

**Comment by rawagner** [comment]
@christianvogt thanks for trying it out!

The easiest scenario would probably be 
1. checkout this PR https://github.com/openshift/console/pull/2601 which changes the behaviour of Inventory card. Currently Inventory Card fetches all resources (all pods, all nodes, all pvcs) and passes them to specific inventory items. Which means that if any resources changes, all items are updated. The PR changes the behavior in a way that not card, but item fetches the need resource ie Pod inventory item fetches all pods, Node item fetches all nodes and so on. I was expecting that this would optimize the number of rerenders - pods changes, only pod item is rerendered, however that is not the case as we already know due to behavior of ConnectToState

2. apply this PR on top of 1. and you will see that how it behaves as expected - if pods changes only pod item is rerendered and others are not affected. Its easier to try this on some resource which does not change often, for example try creating and deleting PVC at see that only PVC item is updated. Without this change, all of them would be.

**Comment by rawagner** [comment]
> Tried out the changes. While sitting on the dashboard page it almost looks like the same cards are being updated but visually nothing has changed as before applying the fix.
> I put conditional breakpoints in the new code and indeed it looks to be working and preventing some re-renders. There are also, as mentioned above, the checks for children are failing and component re-renders are triggered although visually there are no changes.
> 
 Im not sure how to optimize `children`, but `reduxes` can be easily done

**Comment by christianvogt** [comment]
Tried out the fix with the above mentioned PR and it was an improvement. I didn't see any side effects.

**Comment by christianvogt** [comment]
@spadgett what would you like to do?

**Pull request merged**